sklearn/preprocessing/discretization.py

```diff
@@ -152,16 +181,45 @@ def fit(self, X, y=None):
         self.offset_ = offset
         self.n_bins_ = self._validate_n_bins(n_features, ignored)
 
-        ptp = np.ptp(X, axis=0)
-        same_min_max = np.where(ptp == 0)[0]
-        if len(same_min_max) > 0:
-            warnings.warn("Features {} are constant and will be replaced with "
-                          "0.".format(", ".join(str(i) for i in same_min_max)))
+        if self.strategy == 'uniform':
+            ptp = np.ptp(X, axis=0)
+            same_min_max = np.where(ptp == 0)[0]
+            if len(same_min_max) > 0:
+                warnings.warn(
+                    "Features {} are constant and will be replaced with "
+                    "0.".format(", ".join(str(i) for i in same_min_max)))
+
+            with np.errstate(divide='ignore', invalid='ignore'):
+                bin_widths = ptp / self.n_bins_
+            bin_widths[ignored] = 0
+
+        elif self.strategy in ('quantile', 'kmeans'):
+            bin_widths = np.zeros(n_features, dtype=object)
+            for jj in range(n_features):
+                if jj in ignored:
+                    bin_widths[jj] = np.array([])
+                    continue
+                col = X[:, jj][:, None]
+
+                if self.strategy == 'quantile':
+                    n_quantiles = self.n_bins_[jj] + 1
+                    qt = QuantileTransformer(n_quantiles=n_quantiles).fit(col)
```

Comment:
This seems overkill. Just use `boundaries = np.percentile(col, n_quantiles + 1)`